### DISCUSSION TITLE
Multi-age Three Mouse Tracking

### DISCUSSION THREAD

**Discussion by heiley8181** [talmolab/sleap#1625] [Help!] [answered]
Hello SLEAP team,

I am currently using SLEAP for tracking the interactions of three mice within a homecage with an overhead point of view in our videos.   As part of our project, we aim to analyze different age groups of mice. I am curious to know if it is feasible to create a model that can be generalized across all age groups, considering the variations in size as they grow.

Also, I am wondering if there are any suggested recording environment for optimal SLEAP performance. Specifically, does the resolution significantly impact the model training process? Our current recordings resemble the following: 

![image](https://github.com/talmolab/sleap/assets/76165836/a9e3998c-c20e-4725-b458-748c8dd34b75)

Furthermore, I am seeking guidance on optimizing our training settings. It seems like our learning curve reaches max plateau limit at approximately 50-70 for centroid and 70-80 for centered instances. I am uncertain if this is typical or if adjustments to our training settings are advisable. 

![image](https://github.com/talmolab/sleap/assets/76165836/d223e635-c9fe-4dc1-9b3f-123fb54741e5)
![image](https://github.com/talmolab/sleap/assets/76165836/c3aed6eb-ad92-4e20-adb4-84c38a048f43)
![image](https://github.com/talmolab/sleap/assets/76165836/fc5e5cfb-01ef-4482-ad0a-f8bca11de349)

Lastly, I have a general inquiry regarding occlusion. In the context of three mice interactions, numerous instances lead to partial obstruction of body parts, making it challenging for both us and the computer to observe. As I add and label instances, I attempt to anticipate the occluded areas, but it proves to be intricate at times. Right now, I consistently add instances to situations where one mouse remains beneath another for an extended period. Are there any existing features or planned updates that address this occlusion issue in future releases?

Thank you so much in advance!

**Comment by talmo** [comment]
Hi @heiley8181,

Thanks for reaching out! We'll answer your questions inline below.

> Hello SLEAP team,
> 
> I am currently using SLEAP for tracking the interactions of three mice within a homecage with an overhead point of view in our videos. As part of our project, we aim to analyze different age groups of mice. I am curious to know if it is feasible to create a model that can be generalized across all age groups, considering the variations in size as they grow.

Yes, it's definitely possible, but it'll just require a bit more labeling to make sure that you have enough examples in your labels that are representative of all ages/sizes. You'll want to make sure that the model has a large enough receptive field size for the largest animals, but otherwise it should just work across all of them.

You _may_ find that you get better performance by training specialist models for each age, but if they're co-housed then you probably just need to go with a single unified model.

> 
> Also, I am wondering if there are any suggested recording environment for optimal SLEAP performance. Specifically, does the resolution significantly impact the model training process? Our current recordings resemble the following:
> 
> ![image](https://private-user-images.githubusercontent.com/76165836/287960830-a9e3998c-c20e-4725-b458-748c8dd34b75.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3OTg3ODYsIm5iZiI6MTcwMTc5ODQ4NiwicGF0aCI6Ii83NjE2NTgzNi8yODc5NjA4MzAtYTllMzk5OGMtYzIwZS00NzI1LWI0NTgtNzQ4YzhkZDM0Yjc1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDE3NDgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhNTQ2YzliZjViMjA2ZTVmYjAwOTQ1MmU3MmY0MGVlM2FjYTUzOTQ2NWY0NmY0MDVhYTUwZGY4NDczN2U3MzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0rkEhxvRYjdVl9VmGqsc1StAVNrGV32xuA_nJ7Y4fdg)


In terms of your imaging setup, one thing I'd recommend is adding a bit more (or stronger) lighting to boost the contrast a bit, both between animals and against the background.

From the image you provided it seems as if its resolution should not impact performance. The rule of thumb is that you want the finest feature you want to track, to be at least ~2-4 px in diameter. If you're tracking tails, that would be the tail, otherwise the snout if you're ignoring tails. In either case, [based on our previous work](https://www.nature.com/articles/s41592-022-01426-1#:~:text=The%20%E2%80%98mice_of%E2%80%99%20dataset,at%2080%E2%80%89FPS.), you'll want to shoot for ~2 px / mm as your target resolution for mice.


> 
> Furthermore, I am seeking guidance on optimizing our training settings. It seems like our learning curve reaches max plateau limit at approximately 50-70 for centroid and 70-80 for centered instances. I am uncertain if this is typical or if adjustments to our training settings are advisable.
> 
> ![image](https://private-user-images.githubusercontent.com/76165836/287956628-d223e635-c9fe-4dc1-9b3f-123fb54741e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3OTg3ODYsIm5iZiI6MTcwMTc5ODQ4NiwicGF0aCI6Ii83NjE2NTgzNi8yODc5NTY2MjgtZDIyM2U2MzUtYzlmZS00ZGMxLTliM2YtMTIzZmI1NDc0MWU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDE3NDgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQyZmJhYzNkOGZhN2M2OTc3MGM4YTMzODY3ZjEzMzBmZGU3Y2JhNDY0MDQ2Mjg3NDFlMzIxMWU2YTliMjM0MTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jud-hmdozM72QkL5mu_sG5WuotaIYVvsuXuYGCAY9wY) ![image](https://private-user-images.githubusercontent.com/76165836/287956755-c3aed6eb-ad92-4e20-adb4-84c38a048f43.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3OTg3ODYsIm5iZiI6MTcwMTc5ODQ4NiwicGF0aCI6Ii83NjE2NTgzNi8yODc5NTY3NTUtYzNhZWQ2ZWItYWQ5Mi00ZTIwLWFkYjQtODRjMzhhMDQ4ZjQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDE3NDgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhNjkxZDM2OWU0OWY3YmJlNDIzM2Q3ZDI1ZjYzY2RlMzBkNjJjYTliZTJiZjBmNTA3ZWUwYjI2NGU4ZWQ3ZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.L1ibhOtVoQIcsMZ5lzQ3_C02_60K0MFkp_6p1CSw3Aw) ![image](https://private-user-images.githubusercontent.com/76165836/287956822-fc5e5cfb-01ef-4482-ad0a-f8bca11de349.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3OTg3ODYsIm5iZiI6MTcwMTc5ODQ4NiwicGF0aCI6Ii83NjE2NTgzNi8yODc5NTY4MjItZmM1ZTVjZmItMDFlZi00NDgyLWFkMGEtZjhiY2ExMWRlMzQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA1VDE3NDgwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5NDJkODYwMzgxZjZhZmNmYTI0ODM2NTBhYjgwZjk2MDFjZDVhZGQ4MTQxMDViY2Y0NDFlMGZmZjA2YWFlMzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.M11wOOjqV2XQzT6sA19v7MwEM3g5zcHb2BHmTAt1Gko)
> 

It is recommended to have a max receptive field size that is approximately the size of the animal in your frames, to do so you can increase the max stride (under model) to increase your receptive field. This is for your centered instance model (the centroid looks fine).

Alternatively,  you can select the "multi-animal bottom-up" approach under your training pipeline type to see if this provides better results, which we often find to be the case for mouse data.


> Lastly, I have a general inquiry regarding occlusion. In the context of three mice interactions, numerous instances lead to partial obstruction of body parts, making it challenging for both us and the computer to observe. As I add and label instances, I attempt to anticipate the occluded areas, but it proves to be intricate at times. Right now, I consistently add instances to situations where one mouse remains beneath another for an extended period. Are there any existing features or planned updates that address this occlusion issue in future releases?

This is definitely a challenge, especially if it's difficult for you guys to resolve the animals while you're annotating. (In general, if it's hard for you to track something with your eyes, it'll be difficult to train SLEAP to do so as well.)

In your particular case, there are two challenges: (1) getting reliable pose estimates; and (2) reliably tracking identities across frames.

For (1), you can refine your models by selecting more frames that are representative of these instances where animals are overlapping and provide labeling for those frames as best as possible. This may be limited, however, by the imaging conditions as mentioned earlier.

For (2), you'll get drastically improved identity tracking by minimizing the pose estimation errors in (1), so addressing that will immediately help with this step. There are a few avenues to explore in addition to that or if you feel that you've improved the pose estimation as much as you can:

Add visually distinguishing markers to each animal. Common options include shaving parts of the fur in different patterns, dying the fur with nontoxic dyes, or attaching physical markers like collars, ear tags, or tail markings. The best choice depends on the details of your experimental setup and what's technically feasible for you.

Once you have visually distinguishing markers, try using a supervised identity model (they're in the dropdown in the training dialog). These require that you have all your instances set to a consistent track, for example, "dyed mouse" or "shaved mouse". SLEAP will then learn to predict these "identity classes" in addition to the pose. This will rely strictly on the visual appearance markers and basically bypass issues related to identity tracking entirely since each frame is classified independently.

Most of all, I'd recommend improving the poses as much as you can first before resorting to these more invasive approaches.

As far as future updates: we're working on a deep learning-based identity tracking approach that's been showing promising improvements, but it's not quite ready (timeline is a few months). This method will make SLEAP more robust to issues in identity tracking across time, but will come with other trade-offs. It might improve things for your particular application, but it's likely going to be limited by the quality of the initial poses to begin with.

> 
> Thank you so much in advance!

Let us know if you have any other questions!

Cheers,

Talmo

**Reply by heiley8181** [reply]
Hello Talmo,

Thank you so much for your prompt response! I've successfully merged the data from a 45-week project and a 3-week mice project to create a unified model. The model has been generating predictions on a video of 45 weeks. However, I've observed an issue with our model when it comes to capturing the centroid.

To gain more insights, I attempted to analyze the model metrics, but I'm having difficulty interpreting them specifically for the centroid. Below are the model metrics:

![image](https://github.com/talmolab/sleap/assets/76165836/d6568f65-b6d5-40fc-b753-9c6931e84ba9)

![image](https://github.com/talmolab/sleap/assets/76165836/8173ced0-1342-42a7-8bfa-91bda344ad4e)

While the training hyperparameters remained the same as before, I did increase the max stride of the centered instance from 32 to 64. I would greatly appreciate it if you could provide guidance on adjustments that could enhance the performance of our centroid model.

Also, I plan to add a variety of videos from different age groups into a single project. Do you recommend selecting "resume training" each time I add a new video, or would it be more advisable to start the training process from scratch?

Thank you in advance for your help! :)

**Reply by talmo** [reply, marked as answer]
Hi @heiley8181,

So it actually looks like the centroid model is performing pretty much perfectly on your data! Not much to analyze there since the metrics are all nearly perfect. This happens a often with centroid models since that task is not particularly difficult.

The centered instance model could probably improve a bit with respect to the long tail of outliers (especially the literal tail nodes it seems), but overall it's already doing quite well. I'd just recommend adding a few more frames where the model still gets it wrong to keep improving on the edge cases, but otherwise the metrics are looking good.

Max stride of 64: Great!

Adding more videos into single project: Great! I'd advise starting from scratch, but you're welcome to experiment. Since the "Resume Training" option depends entirely on where you're starting from, it's hard to say exactly how it'll perform (sometimes it gets stuck in a local minimum due to biases in the previous training data), but in theory it should save a bit of time. If you have the time though, just start it from scratch to get a nice and more reproducible result.

Let us know if you have any questions!

Cheers,

Talmo
